src/lib/replication/repl_dev/raft_repl_dev.cpp

```diff
@@ -219,9 +313,149 @@ repl_req_ptr_t RaftReplDev::follower_create_req(repl_key const& rkey, sisl::blob
+                                             rreq->remote_blkid.server_id /* blkid_originator */,
+                                             builder->CreateVector(rreq->remote_blkid.blkid.serialize().cbytes(),
+                                                                   rreq->remote_blkid.blkid.serialized_size())));
+        LOGINFO("Fetching data from remote: rreq=[{}], remote_blkid={}", rreq->to_compact_string(),
+                rreq->remote_blkid.blkid.to_string());
+    }
+
+    builder->FinishSizePrefixed(
+        CreateFetchData(*builder, CreateFetchDataRequest(*builder, builder->CreateVector(entries))));
+
+    // leader can change, on the receiving side, we need to check if the leader is still the one who originated the
+    // blkid;
+    group_msg_service()
+        ->data_service_request_bidirectional(
+            nuraft_mesg::role_regex::LEADER, FETCH_DATA,
+            sisl::io_blob_list_t{
+                sisl::io_blob{builder->GetBufferPointer(), builder->GetSize(), false /* is_aligned */}})
+        .via(&folly::InlineExecutor::instance())
+        .thenValue([this, builder, rreqs](auto e) {
+            RD_REL_ASSERT(!!e, "Error in fetching data");
+
+            auto raw_data = e.value().cbytes();
+            auto total_size = e.value().size();
+
+            for (auto const& rreq : *rreqs) {
+                auto const data_size = rreq->remote_blkid.blkid.blk_count() * get_blk_size();
+                // if data is already received, skip it because someone is already doing the write;
+                if (rreq->state.load() & uint32_cast(repl_req_state_t::DATA_RECEIVED)) {
+                    // very unlikely to arrive here, but if data got received during we fetch, let the data channel
+                    // handle data written;
+                    raw_data += data_size;
+                    total_size -= data_size;
+
+                    // if blk is already allocated, validate if blk is valid and size matches;
+                    RD_DBG_ASSERT(rreq->local_blkid.is_valid(), "Invalid blkid for rreq={}", rreq->to_compact_string());
+                    auto const local_size = rreq->local_blkid.blk_count() * get_blk_size();
+                    RD_DBG_ASSERT_EQ(data_size, local_size,
+                                     "Data size mismatch for rreq={} blkid={}, remote size: {}, local size: {}",
+                                     rreq->to_compact_string(), rreq->local_blkid.to_string(), data_size, local_size);
+
+                    RD_LOG(INFO, "Data Channel: Data already received for rreq=[{}], skip and move on to next rreq.",
+                           rreq->to_compact_string());
+                    continue;
+                } else {
+                    // aquire lock here to avoid two threads are trying to do the same thing;
+                    std::unique_lock< std::mutex > lg(rreq->state_mtx);
```

Comment:
thenValue will be executed in a single thread, right? and after each log batch append, the request are also different.
so , do we really the lock here?

Reply:
you are right about here in this context. The lock is to protect from raft log channel and data channel that could also be allocating blks (in case for some reason data arrives after this fetch remote is sent)